Framework/PythonInterface/plugins/algorithms/FindGlobalBMatrix.py

```diff
@@ -0,0 +1,263 @@
+        self.declareProperty(name="alpha", defaultValue=-1.0,
+                             direction=Direction.Input,
+                             validator=angleValidator,
+                             doc="Lattice angle alpha")
+        self.declareProperty(name="beta", defaultValue=-1.0,
+                             direction=Direction.Input,
+                             validator=angleValidator,
+                             doc="Lattice angle beta")
+        self.declareProperty(name="gamma", defaultValue=-1.0,
+                             direction=Direction.Input,
+                             validator=angleValidator,
+                             doc="Lattice angle gamma")
+        self.declareProperty(name="Tolerance", defaultValue=0.15,
+                             direction=Direction.Input,
+                             validator=positiveFloatValidator,
+                             doc="Tolerance to index peaks in in H,K and L")
+
+    def validateInputs(self):
+        issues = dict()
+        ws_list = self.getProperty("PeakWorkspaces").value
+        n_valid_ws = 0
+        for wsname in ws_list:
+            ws = AnalysisDataService.retrieve(wsname)
+            if isinstance(ws, IPeaksWorkspace) and ws.getNumberPeaks() >= _MIN_NUM_PEAKS:
+                n_valid_ws += 1
+        if n_valid_ws < 2 or n_valid_ws < len(ws_list):
+            issues["PeakWorkspaces"] = f"Accept only peaks workspace with more than {_MIN_NUM_PEAKS} peaks - " \
+                                       "there must be at least two peak tables provided in total."
+        return issues
+
+    def PyExec(self):
+        # setup progress bar
+        prog_reporter = Progress(self, start=0.0, end=1.0, nreports=3)
+        # Get input
+        ws_list = self.getProperty("PeakWorkspaces").value
+        a = self.getProperty('a').value
+        b = self.getProperty('b').value
+        c = self.getProperty('c').value
+        alpha = self.getProperty('alpha').value
+        beta = self.getProperty('beta').value
+        gamma = self.getProperty('gamma').value
+        self.tol = self.getProperty('Tolerance').value
+
+        # Find initial UB and use to index peaks in all runs
+        prog_reporter.report(1, "Find initial UB for peak indexing")
+        self.find_initial_indexing(a, b, c, alpha, beta, gamma, ws_list)  # removes runs from ws_list if can't index
+
+        # optimize the lattice parameters across runs (i.e. B matrix)
+        prog_reporter.report(2, "Optimize B")
+
+        def fobj(x):
+            return self.calcResiduals(x, ws_list)
+
+        alatt0 = [a, b, c, alpha, beta, gamma]
+        alatt, cov, info, msg, ier = leastsq(fobj, x0=alatt0, full_output=True)
+        success = ier in [1, 2, 3, 4]
+        if success:
+            # calculate errors
+            dof = sum([self.child_IndexPeaks(ws, RoundHKLs=True) for ws in ws_list]) - len(alatt0)
+            err = np.sqrt(abs(np.diag(cov)) * (info['fvec'] ** 2).sum() / dof)
```

Comment:
Again by putting daft parameters into the FindGlobalBMatrix call I occasionally get an error from this line because "cov" is set to None:

![image](https://user-images.githubusercontent.com/56295817/126135312-b3f47146-ac3f-425d-9ea0-a4434984f4fd.png)

I think the code should check here if cov is not None before trying to calculate err.

In case it helps, the other return parameters from the call to leastsq in this situation are:
ier=2
msg = 'The relative error between two consecutive iterates is at most 0.000000'
alatt and info are populated with what looks like sensible data

Reply:
Have added this check and a comment to explain that the cov matrix is set to None is singular (but fit can still be a success)